boot/boot_test.go

```diff
@@ -902,3 +902,457 @@ func (s *bootenv20Suite) TestMarkBootSuccessful20BaseUpdate(c *C) {
+// SetBootVars, for the kernel snap when we commit the boot state during
+// MarkBootSuccessful
+func (s *bootenv20Suite) TestHappyMarkBootSuccessfulKernelUpgradeSetBootVarsPanics(c *C) {
+	r := boottest.ForceModeenv(dirs.GlobalRootDir, &boot.Modeenv{
+		Mode:           "run",
+		RecoverySystem: "20191018",
+		Base:           "core20_1.snap",
+	})
+	defer r()
+
+	coreDev := boottest.MockUC20Device("some-snap")
+	c.Assert(coreDev.HasModeenv(), Equals, true)
+
+	// we are trying a new kernel snap
+	s.bootloader.BootVars["kernel_status"] = boot.TryingStatus
+
+	// set the current Kernel
+	kernel1, err := snap.ParsePlaceInfoFromSnapFileName("pc-kernel_1.snap")
+	c.Assert(err, IsNil)
+	r = s.bootloader.SetRunKernelImageEnabledKernel(kernel1)
+	defer r()
+
+	// set the current try kernel
+	kernel2, err := snap.ParsePlaceInfoFromSnapFileName("pc-kernel_2.snap")
+	c.Assert(err, IsNil)
+	r = s.bootloader.SetRunKernelImageEnabledTryKernel(kernel2)
+	defer r()
+
+	// we will panic during the next SetBootVars, which is the first step, so
+	// essentially we reboot before anything is committed
+	restoreBootloaderPanic := s.bootloader.SetRunKernelImagePanic("SetBootVars")
+
+	// attempt mark successful, expecting a panic in SetBootVars
+	c.Check(
+		func() { boot.MarkBootSuccessful(coreDev) },
+		PanicMatches,
+		"mocked reboot panic in SetBootVars",
+	)
+
+	// don't panic anymore
+	restoreBootloaderPanic()
+
+	// do the bootloader kernel failover logic handling
+	nextBootingKernel, err := runBootloaderLogic(c, s.bootloader)
+	c.Assert(err, IsNil)
+
+	// the next kernel we boot from should be kernel1, because we didn't
+	// actually commit anything
+	c.Assert(nextBootingKernel, Equals, kernel1)
+
+	// kernel_status should still be default
+	m, err := s.bootloader.GetBootVars("kernel_status")
+	c.Assert(err, IsNil)
+	c.Assert(m, DeepEquals, map[string]string{"kernel_status": boot.DefaultStatus})
+
+	// try again, should complete now
+	err = boot.MarkBootSuccessful(coreDev)
+	c.Assert(err, IsNil)
+
+	// after running now, we should still have Kernel() == pc-kernel_1.snap
+	// because the try failed
+
+	afterKernel, err := s.bootloader.Kernel()
+	c.Assert(err, IsNil)
+	c.Assert(afterKernel, DeepEquals, kernel1)
+
+	m, err = s.bootloader.GetBootVars("kernel_status")
+	c.Assert(err, IsNil)
+	c.Assert(m, DeepEquals, map[string]string{"kernel_status": boot.DefaultStatus})
+
+	// we should also still have TryKernel() == pc-kernel_2.snap, as well as
+	// trying == true, because the symlink is still there, the snapstate mgr
+	// is responsible for cleaning it up when it notices the boot with the try
+	// kernel failed - that is done a level above this test
+	afterTryKernel, err := s.bootloader.TryKernel()
+	c.Assert(err, IsNil)
+	c.Assert(afterTryKernel, Equals, kernel2)
+}
+
+// TestHappyMarkBootSuccessfulKernelUpgradeEnableKernelPanics emulates a reboot during
+// EnableKernel, for the kernel snap when we commit the boot state during
+// MarkBootSuccessful
```

Comment:
do the tests cover all the possible interesting reboot points? or just some?  I wonder if there's a way to write some of these more in a tabular style

Reply:
Yes to my knowledge it does cover all the "interesting" reboot points for kernel snaps specifically. The only cases we don't currently cover are situations where a subset of the critical operations are called, i.e. when calling SetNextBoot on the current kernel snap, where we only call SetBootVars. I can add these, but it's not "interesting" since there's only one critical operation we do, which is setting boot variables.

To be clear, setting aside the uninteresting cases above, we have the following sequences of critical operations:

### MakeBootable
- SetBootVars
- EnableKernel
- DisableTryKernel
- Modeenv.Write

### SetNextBoot
- Modeenv.Write
- EnableTryKernel
- SetBootVars

Setting aside reboots during/before the Modeenv.Write, we have these tests:

### MakeBootable
- TestHappyMarkBootSuccessfulKernelSetBootVarsPanics (note this should be renamed, but as per your other comment they should all be renamed)
- TestHappyMarkBootSuccessfulKernelUpgradeEnableKernelPanics
- TestHappyMarkBootSuccessfulKernelUpgradeDisableTryKernelPanics

### SetNextBoot
- TestHappyCoreParticipant20SetNextKernelSnapEnableTryKernelPanics
- TestHappyCoreParticipant20SetNextKernelSnapSetBootVarsPanics


I would like to add the Modeenv.Write reboots to a future PR where we move the snap-bootstrap logic into boot pkg so we can consume that here more easily, but I could probably do that here without needing that logic I guess. But we will need that logic moved to be able to effectively write base snap tests like this however, which I think should be a different PR.

Finally, it could be done in a tabular way but it gets messy I think since there's so many things to check. If you want I can change it like that however.

Reply:
thanks for the answer, it seems the current status of things is good, and we can rethink later